homeassistant/helpers/restore_state.py

```diff
@@ -2,97 +2,145 @@
-from homeassistant.loader import bind_hass
-from homeassistant.components.history import get_states, last_recorder_run
-from homeassistant.components.recorder import (
-    wait_connection_ready, DOMAIN as _RECORDER)
-import homeassistant.util.dt as dt_util
+DATA_RESTORE_STATE_TASK = 'restore_state_task'
 
-RECORDER_TIMEOUT = 10
-DATA_RESTORE_CACHE = 'restore_state_cache'
-_LOCK = 'restore_lock'
 _LOGGER = logging.getLogger(__name__)
 
+STORAGE_KEY = 'core.restore_state'
+STORAGE_VERSION = 1
+
+# How long between periodically saving the current states to disk
+STATE_DUMP_INTERVAL = timedelta(minutes=15)
+
+
+class RestoreStateData():
+    """Helper class for managing the helper saved data."""
+
+    @staticmethod
+    async def async_get_instance(hass: HomeAssistant) -> 'RestoreStateData':
+        """Get the singleton instance of this data helper."""
+        task = hass.data.get(DATA_RESTORE_STATE_TASK)
+
+        if task is None:
+            async def load_instance(hass: HomeAssistant) -> 'RestoreStateData':
+                """Set up the restore state helper."""
+                data = RestoreStateData(hass)
+
+                try:
+                    states = await data.store.async_load()
+                except HomeAssistantError as exc:
+                    _LOGGER.error("Error loading last states", exc_info=exc)
+                    states = None
+
+                if states is None:
+                    _LOGGER.debug('Not creating cache - no saved states found')
+                    data.last_states = {}
+                else:
+                    data.last_states = {
+                        state['entity_id']: State.from_dict(state)
+                        for state in states}
+                    _LOGGER.debug(
+                        'Created cache with %s', list(data.last_states))
+
+                if hass.state == CoreState.running:
+                    data.async_setup_dump()
+                else:
+                    hass.bus.async_listen_once(
+                        EVENT_HOMEASSISTANT_START, data.async_setup_dump)
+
+                return data
+
+            task = hass.data[DATA_RESTORE_STATE_TASK] = hass.async_create_task(
+                load_instance(hass))
+
+        return await task
+
+    def __init__(self, hass: HomeAssistant) -> None:
+        """Initialize the restore state data class."""
+        self.hass = hass  # type: HomeAssistant
+        self.store = Store(hass, STORAGE_VERSION, STORAGE_KEY,
+                           encoder=JSONEncoder)  # type: Store
+        self.last_states = {}  # type: Dict[str, State]
+        self.entities = []  # type: List[RestoreEntity]
+
+    async def async_dump_states(self) -> None:
+        """Save the current state machine to storage."""
+        _LOGGER.debug("Dumping states")
+        # Entity ID set of registered restorable entities to dump
+        entity_ids = set(entity.entity_id for entity in self.entities)
```

Comment:
Any reason we don't just make `self.entities` a set to begin with?

Reply:
Entities don't implement `__hash__()`, but based on Martin's advice, we're just going to store entity_ids instead of objects, so we'll be able to simplify this.